humility-core/src/arch/rv.rs

```diff
@@ -72,6 +72,27 @@ impl Arch for RVArch {
         RVRegister::iter().map(Register::RiscV).collect()
     }
 
+    //
+    // TODO: will first check for `CURRENT_TASK_PTR` then check mscratch/sscratch,
+    // When using xscratch, there is a small time when the current task ptr is
+    // actually in a0 when handling a trap, we can detect this and read the
+    // correct register.
+    //
+    fn get_current_task_ptr(
+        &self,
+        hubris: &HubrisArchive,
+        core: &mut dyn crate::core::Core,
+    ) -> Result<u64> {
+        match hubris.lookup_symword("CURRENT_TASK_PTR") {
+            Ok(ptr) => Ok(core.read_word_32(ptr)? as u64),
+            // Means current task is in mscratch or sscratch, but only if on riscv
+            Err(_) => {
+                // TODO right now blindly check mscratch, but should check that it is valid
+                core.read_reg(Register::RiscV(RVRegister::MSCRATCH))
```

Comment:
How do you intend to determine whether mscratch/sscratch is appropriate to read?

Reply:
See https://rivosinc.atlassian.net/browse/SW-600 for where @furquan-rivos and I discussed it some.  Within the trap handler we temporarily move the task ptr from scratch to a0.  I should be able to extract the address range where this is happens by looking at the register transfers that occur in _start_trap.  Then compare against PC and read either `a0` or scratch.